tests/unit/runners/test_vault.py

```diff
@@ -364,3 +364,175 @@ def test_generate_token(self):
                 call("http://fake_url", headers=ANY, json=ANY, verify=ANY),
             ]
             mock.assert_has_calls(calls)
+
+
+class VaultAWSIAMRoleAuthTest(TestCase, LoaderModuleMockMixin):
+    """
+    Tests for the runner module of the Vault with approle setup
+    """
+
+    def setup_loader_modules(self):
+        return {
+            vault: {
+                "__opts__": {
+                    "vault": {
+                        "url": "http://127.0.0.1",
+                        "auth": {"method": "aws", "role": "test", "aws_method": "iam"},
+                    }
+                }
+            }
+        }
+
+    @patch("salt.runners.vault._validate_signature", MagicMock(return_value=None))
+    @patch(
+        "salt.runners.vault._get_token_create_url",
+        MagicMock(return_value="http://fake_url"),
+    )
+    @patch(
+        "salt.utils.aws.creds",
+        MagicMock(
+            return_value=(
+                "AKIAIOSFODNN7EXAMPLE",
+                "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
+                "AQoEXAMPLEH4ao",
+            )
+        ),
+    )
+    def test_generate_token(self):
+        """
+        Basic test for test_generate_token with approle (two vault calls)
+        """
+        mock = _mock_json_response(
+            {"auth": {"client_token": "test", "renewable": False, "lease_duration": 0}}
+        )
+        with patch("requests.post", mock):
+            result = vault.generate_token("test-minion", "signature")
+            log.debug("generate_token result: %s", result)
+            self.assertTrue(isinstance(result, dict))
+            self.assertFalse("error" in result)
+            self.assertTrue("token" in result)
+            self.assertEqual(result["token"], "test")
+            calls = [
+                call(
+                    "http://127.0.0.1/v1/auth/aws/login",
+                    headers=ANY,
+                    json=ANY,
+                    verify=ANY,
+                ),
+                call("http://fake_url", headers=ANY, json=ANY, verify=ANY),
+            ]
+            mock.assert_has_calls(calls)
+
+
+class VaultAWSEC2RoleAuthTest(TestCase, LoaderModuleMockMixin):
+    """
+    Tests for the runner module of the Vault with approle setup
+    """
+
+    def setup_loader_modules(self):
+        return {
+            vault: {
+                "__opts__": {
+                    "vault": {
+                        "url": "http://127.0.0.1",
+                        "auth": {"method": "aws", "role": "test", "aws_method": "ec2"},
+                    }
+                }
+            }
+        }
+
+    @patch("salt.runners.vault._validate_signature", MagicMock(return_value=None))
```

Comment:
**suggestion**Patches already create mocks. Also it's **usually** preferable to use `autospec=True`:

```
@patch("salt.runners.vault._validate_signature", autospec=True, return_value=None)
```

This suggestion is broadly applicable.

Reply:
Hello can you please explain me a bit more not sure to understand it completly.
To write new tests I just take the same approach of upstream. 
[https://github.com/saltstack/salt/blob/094687af0565c2e410b77e3d35de3c9807dd404c/tests/unit/runners/test_vault.py#L338](url) doesn't use `autospec=True` neither.

Should we change upstream testcase as well ? :/

Reply:
Hello.
Can you have a look to my previous comment ? thank you

Reply:
Hey @jynolen sorry for the extremely late reply, apparently this PR got lost in my GitHub noise :disappointed: 

`autospec=True` is almost always a good thing, because it will match the function signatures. For example:

```
def add(a, b):
    ...
    
with patch("__main__.add", return_value=42):
    fake_add(3, 4, 10, 100)
```

That will work, because MagicMock's are extraordinarily flexible. But with `autospec=True` you'd get a `TypeError: too many positional arguments`.

There are some exceptions and reasons you might not want to autospec, but if you're making a mock then almost always autospec is the right way to go.

I would say that it's not *necessary* to change the other tests, but I would definitely approve of changing the other tests (assuming that they're not one of the exceptions :joy:)